### PR TITLE
Clarify some parts of Application Messages section.

### LINKED ISSUE #679: Content encryption: Clarify generation sequence
The generation numbers should be increased sequentially, gaps should be avoided to avoid unnecessary expensive operations on clients. Clients might also abort if the gap is too large to prevent attacks from malicious insiders.

### PR DESCRIPTION
**Pull request by Bren2010**
Closes #679

### CODE REVIEW

File: draft-ietf-mls-protocol.md [hunk 1]
```diff
@@ -4488,16 +4471,20 @@ state compromise.
...
+are received out of order, the client moves the sender ratchet forward to match
+the received generation counter. Any unused nonce and key pairs from the ratchet
+are potentially stored so that they can be used to decrypt the messages which
+were delayed or reordered.
+
+Applications SHOULD define a policy on how long to keep unused nonce and key
+pairs for a sender, and the maximum number to keep. This is in addition to
+ensuring that these nonce and key pairs are promptly deleted when the epoch
+ends. Applications SHOULD also define a policy limiting the maximum number of
+steps that clients will move the ratchet forward in one step. Messages received
```
Comment (bifurcation) [inline draft-ietf-mls-protocol.md:4485]:
```suggestion
steps that clients will move the ratchet forward in response to a new message. Messages received
```

File: draft-ietf-mls-protocol.md [hunk 2]
```diff
@@ -4488,16 +4471,20 @@ state compromise.
...
+with an unusually high generation counter would then be rejected to avoid
+causing a denial-of-service attack.
```
Comment (bifurcation) [inline draft-ietf-mls-protocol.md:4487]:
```suggestion
causing a denial-of-service attack by requiring the recipient to perform an
excessive number of key derivations.  For example, a malicious group member
could send a message with `generation = 0xffffffff` at the beginning of a new 
epoch, forcing recipients to perform billions of key derivations.
```

**Review by bifurcation — APPROVED** [review]
Couple of friendly amendments, otherwise lgtm.

I am going to hold this as slightly more than editorial, given the number of MUSTs that are affected.  But I would note for others that the MUSTs that are removed here are all redundant with requirements above.

File: draft-ietf-mls-protocol.md [hunk 3]
```diff
@@ -4488,16 +4471,23 @@ state compromise.
...
+
+Applications SHOULD define a policy on how long to keep unused nonce and key
+pairs for a sender, and the maximum number to keep. This is in addition to
+ensuring that these nonce and key pairs are promptly deleted when the epoch
+ends. Applications SHOULD also define a policy limiting the maximum number of
+steps that clients will move the ratchet forward in response to a new message. Messages received
+with an unusually high generation counter would then be rejected to avoid
+causing a denial-of-service attack by requiring the recipient to perform an
+excessive number of key derivations.  For example, a malicious group member
+could send a message with `generation = 0xffffffff` at the beginning of a new
```
Comment (Bren2010) [inline draft-ietf-mls-protocol.md:4489]:
```suggestion
could send a message with `generation = 0xffffffff` at the beginning of a new
```

File: draft-ietf-mls-protocol.md [hunk 4]
```diff
@@ -4488,16 +4471,23 @@ state compromise.
...
+the received generation counter. Any unused nonce and key pairs from the ratchet
+are potentially stored so that they can be used to decrypt the messages which
+were delayed or reordered.
+
+Applications SHOULD define a policy on how long to keep unused nonce and key
+pairs for a sender, and the maximum number to keep. This is in addition to
+ensuring that these nonce and key pairs are promptly deleted when the epoch
+ends. Applications SHOULD also define a policy limiting the maximum number of
+steps that clients will move the ratchet forward in response to a new message. Messages received
+with an unusually high generation counter would then be rejected to avoid
```
Comment (raphaelrobert) [inline draft-ietf-mls-protocol.md:4486]: "unusually high" is a bit vague. Maybe we can be more specific here by saying that clients can enforce a maximum difference for the generation counter between two subsequent messages.

**Review by raphaelrobert — APPROVED** [review]
Thanks for picking this up! Just one nit, but looks good otherwise.

**Comment by bifurcation** [comment]
Interim 2022-05-26:

* Intended to be mostly editorial
* Leaving open for now
* Clear to merge in between meetings if folks are comfortable

**Comment by bifurcation** [comment]
Interim 2022-06-02:

* Still clear from last week

**Pull request merged**